nesis/api/tests/core/document_loaders/test_sharepoint.py

```diff
@@ -0,0 +1,171 @@
+        "engine": "sharepoint",
+        "connection": {
+            "site_url": "https://ametnes.sharepoint.com/sites/nesis-test/",
+            "client_id": "<sharepoint_app_client_id>",
+            "tenant": "<sharepoint_app_tenant_id>",
+            "thumbprint": "<sharepoint_app_cert_thumbprint>",
+            "certificate_path": "path_to_private_cert_keyfile",
+            "file_locations": "Shared Documents, Contracts",
+        },
+    }
+
+    datasource = Datasource(
+        name=data["name"],
+        connection=data["connection"],
+        source_type=DatasourceType.SHAREPOINT,
+        status=DatasourceStatus.ONLINE,
+    )
+
+    session.add(datasource)
+    session.commit()
+
+    sp_client_context = mock.MagicMock()
+    document_library = mock.MagicMock()
+    item = mock.MagicMock()
+    file_mock = mock.MagicMock()
+    _items = mock.MagicMock()
+
+    client_context().with_client_certificate.return_value = sp_client_context
+
+    sp_client_context.web.default_document_library.return_value = document_library
+
+    type(file_mock).name = mock.PropertyMock(return_value="some_file.pdf")
+    type(file_mock).serverRelativeUrl = mock.PropertyMock(
+        return_value=r"/sites/nesit-test/Shared Documents/some_file.pdf"
+    )
+    type(file_mock).unique_id = mock.PropertyMock(return_value="123123-123131-1313")
+    type(file_mock).last_time_last_modified = mock.PropertyMock(
+        return_value=datetime.time()
+    )
+    type(file_mock).length = mock.PropertyMock(return_value=1023)
+    type(file_mock).author = mock.PropertyMock(return_value="file author")
+
+    type(item).file_system_object_type = mock.PropertyMock(return_value=0)
+    type(item).file = mock.PropertyMock(return_value=file_mock)
+    document_library.items = _items
+
+    document_library.items.select.return_value.expand.return_value.get_all.return_value.execute_query.return_value = [
+        item
+    ]
+
+    file_download = mock.MagicMock()
+    file_mock.download.return_value.execute_query.return_value = file_download
+
+    http_client = mock.MagicMock()
+    http_client.upload.return_value = json.dumps({})
+
+    sharepoint.fetch_documents(
+        connection=data["connection"],
+        http_client=http_client,
+        metadata={"datasource": "documents"},
+        rag_endpoint="http://localhost:8080",
+        cache_client=cache,
+    )
+
+    _, upload_kwargs = http_client.upload.call_args_list[0]
+    url = upload_kwargs["url"]
+    file_path = upload_kwargs["filepath"]
+    metadata = upload_kwargs["metadata"]
+    field = upload_kwargs["field"]
+
+    assert url == f"http://localhost:8080/v1/ingest/files"
+    assert file_path.endswith("some_file.pdf")
+    assert field == "file"
+    ut.TestCase().assertDictEqual(
+        metadata,
+        {
+            "datasource": "documents",
+            "file_name": "some_file.pdf",
+            "self_link": r"/sites/nesit-test/Shared Documents/some_file.pdf",
+        },
+    )
+
```

Comment:
The test above tests the sync (create) document process so we just need another test for the un_sync process and another one for the sync (update). See the test [nesis/api/tests/core/document_loaders/test_samba.py](https://github.com/ametnes/nesis/blob/19-feature-add-sharepoint-datasoure/nesis/api/tests/core/document_loaders/test_samba.py)